### PR TITLE
[tree] Enable TTreeCache even if fAutoFlush == 0

### LINKED ISSUE #8713: [tree] TTreeCache is turned off when `fAutoFlush == 0`
### Describe the bug

Due to the logic in `TTree::GetCacheAutoSize`, if `fAutoFlush == 0` then the size of the `TTreeCache` is set to 0 which in turn disables it.

However there are separate mechanisms to disable the `TTreeCache` if desired (e.g. `TTree::SetCacheSize`) and even if `fAutoFlush == 0`, when reading over the network, we can benefit greatly from `TTreeCache`'s pre-fetching.

### Additional context

- discussion and debugging: https://mattermost.web.cern.ch/root/pl/uzub3wwt3bdxxcdu4jmxxyrmby
- summary: https://mattermost.web.cern.ch/root/pl/xmtuq55j3pnc7cp7f8zyhf5ujw

### PR DESCRIPTION
**Pull request by eguiraud**
With the current logic, if for some reason fAutoFlush is set to 0
for the TTree, the TTreeCache is disabled.
That is undesirable: we still can and want to do pre-fetching even
if auto-flushing was turned off when writing the TTree.

Other more direct methods to turn off the TTreeCache still work,
e.g. tree->SetCacheSize(0).

## Checklist:

- [x] tested changes locally (complicated because of https://github.com/root-project/root/issues/7366 , but seems mostly ok) 
- [ ] updated the docs (@pcanal let me know if there are docs to update)
- [x] could we set the cachesize to a better value than the autoflush default of ~30MB?

This PR fixes #8713 .

### CODE REVIEW

**Comment by pcanal** [comment]
> could we set the cachesize to a better value than the autoflush default of ~30MB?

Yes, you could use the `median` of `fClusterSize` (weighted by the number of cluster, which is encoded in `fClusterRangeEnd`, see `TTree::Print` for usage example) [`median` (or something higher) is better here than the average as we want to the size to cover most cluster (the average is  likely to be slightly too small)]

**Comment by eguiraud** [comment]
The median will still leave many clusters out (it only guarantees that 50% of them will be in, right?). How about the mode?

**Comment by pcanal** [comment]
Yes, I agree that the mode, when well defined would be better (As a aside the median an match more than 50% of the cluster, as is the case in the example file you had [where actually the mode and the median are the same]).

By "well defined", I mean that if a large fraction of the file is not clustered (i.e result of the merge of many unclustered file and a few clustered files) then the 'mode' might not cover the majority of the cluster.   

I would say we could use "the mode or the median which ever cover more clusters and/or the largest fraction of the file", isn't it?

**Comment by eguiraud** [comment]
> the mode or the median which ever cover more clusters and/or the largest fraction of the file

Alright, can do that. but it's not super cheap, so I have two questions.

1. Should I:

a. add a TTree method that evaluates that thing and call it from `TTree::GetCacheAutoSize`
b. add a TTree method that evaluates that thing and add a data member to TTree to cache its value after the first time I compute it
c. evaluate that thing in TTree's constructor and store it in a data member
d. add a free function that takes a TTree and evaluates that thing
e. none of the above, something else

2. What should we do for TChains? Override that calculation so that it errors out?

**Comment by pcanal** [comment]
> What should we do for TChains? Override that calculation so that it errors out?

Humm .. i think it somehow get delegated to the underlying TTree.

> But it's not super cheap, 

it should be a one time (per TTree) cost in the case where the user did not specify the cache size and fAutoFlush is to zero (because after that the CacheSize is set)

> add a TTree method that evaluates that thing and call it from TTree::GetCacheAutoSize

I would do that.

Thanks

**Comment by eguiraud** [comment]
Latest hurdle: `fClusterSize` is a number of entries, the cache size (which we want to set to the median/mode of cluster sizes) has to be in bytes (compressed or uncompressed, I'm not sure) -- how do I go from median/mode of the number of entries in a cluster to the desired size in bytes?

**Comment by eguiraud** [comment]
Now, if `fAutoFlush == 0`, we are setting the cache size to `1.5 * medianClusterSize * GetZipBytes() / (fEntries + 1)` (mimicking the logic in `GetCacheAutoSize` but using the median cluster size instead of `fAutoFlush`). When calculating the median, cluster ranges that report `fClusterSize == 0` are ignored.

File: tree/tree/src/TTree.cxx [hunk 1]
```diff
@@ -8220,6 +8220,41 @@ void TTree::MarkEventCluster()
     ++fNClusterRange;
 }
 
+/// Estimate a reasonable cache size based on the cluster sizes.
+/// This method is meant to be called if all other heuristics fail (see GetCacheAutoSize()).
+Long64_t TTree::EstimateCacheSize()
```
Comment (pcanal) [inline tree/tree/src/TTree.cxx:8225]: `GetCacheAutoSize` intent/purpose is to estimate the [default] cache size :), so this name is a bit confusing.
Consider renaming this function `EstimateClusterSize` and move the final calculation (number of clusters to cache size) into the `GetCacheAutoSize`.

Comment (eguiraud) [reply tree/tree/src/TTree.cxx]: done (i called the new method `GetMedianClusterSize`, split the PR in two commits)

File: tree/tree/src/TTree.cxx [hunk 2]
```diff
@@ -5375,9 +5375,15 @@ Long64_t TTree::GetCacheAutoSize(Bool_t withDefault /* = kFALSE */ ) const
...
    Long64_t cacheSize = 0;
 
-   if (fAutoFlush < 0) cacheSize = Long64_t(-cacheFactor*fAutoFlush);
-   else if (fAutoFlush == 0) cacheSize = 0;
-   else cacheSize = Long64_t(cacheFactor*1.5*fAutoFlush*GetZipBytes()/(fEntries+1));
+   if (fAutoFlush < 0) {
+      cacheSize = Long64_t(-cacheFactor * fAutoFlush);
+   } else if (fAutoFlush == 0) {
+      const auto medianClusterSize = GetMedianClusterSize();
+      // FIXME @philippe should we multiply this by cacheFactor, here and below?
```
Comment (pcanal) [inline tree/tree/src/TTree.cxx:5382]: Yes, indeed.

Comment (eguiraud) [reply tree/tree/src/TTree.cxx]: done (only here, not below to mimic the pre-existing calculations)

File: tree/tree/src/TTree.cxx [hunk 3]
```diff
@@ -5375,9 +5375,17 @@ Long64_t TTree::GetCacheAutoSize(Bool_t withDefault /* = kFALSE */ ) const
...
-   else if (fAutoFlush == 0) cacheSize = 0;
-   else cacheSize = Long64_t(cacheFactor*1.5*fAutoFlush*GetZipBytes()/(fEntries+1));
+   if (fAutoFlush < 0) {
+      cacheSize = Long64_t(-cacheFactor * fAutoFlush);
+   } else if (fAutoFlush == 0) {
+      const auto medianClusterSize = GetMedianClusterSize();
+      if (medianClusterSize > 0)
+         cacheSize = Long64_t(cacheFactor * 1.5 * medianClusterSize * GetZipBytes() / (fEntries + 1));
+      else
+         cacheSize = 30000000; // let's set cacheSize to a reasonable default (the default value of fAutoFlush in bytes)
```
Comment (pcanal) [inline tree/tree/src/TTree.cxx:5385]: Maybe to be consistent:
```suggestion
         cacheSize = Long64_t(cacheFactor * 1.5 * 30000000); // let's set cacheSize to a reasonable default (the default value of fAutoFlush in bytes)
```

Comment (eguiraud) [reply tree/tree/src/TTree.cxx]: done, thanks

**Pull request merged**